leader-election-impl/src/main/java/com/palantir/leader/proxy/AsyncRetrier.java

```diff
@@ -53,20 +54,33 @@
         return execute(supplier, maxAttempts);
     }
 
+    /*
+    Threading model: Given the context of this class, it is likely that the first request will pass. Since this occurs
+    in the hot path, we should probably avoid putting it on another thread to do what would be light work as we will
+    pay for that in context switches.
+
+    In a similar vein to reduce context switches, when the supplier::get call returns, the processing should be fairly
+    lightweight, so we should be able to do it in whatever thread ran/took control of the original supplier::get call,
+    hence the usage of directExecutor.
+
+    If we do have to retry, then we're likely "slow" enough that we should give up the initial calling thread and
+    offload the work onto a separate executor.
+     */
     private ListenableFuture<T> execute(Supplier<ListenableFuture<T>> supplier, int retriesRemaining) {
-        return Futures.transformAsync(Futures.submitAsync(supplier::get, executionExecutor),
+        return Futures.transformAsync(Futures.submitAsync(supplier::get, MoreExecutors.directExecutor()),
```

Comment:
nit: should we remove the inner `transformAsync` as it should be unnecessary if we're getting from the supplier on the current thread?
```suggestion
        return Futures.transformAsync(supplier.get(),
```

Reply:
A good shout, thanks!

Reply:
Had to wrap the supplier so we don't throw but return failed futures, but I think it's fine.